### DISCUSSION TITLE
Issue compiling ESP32 port on Windows

### DISCUSSION THREAD

**Discussion by Lana-chan** [micropython/micropython#14169] [ESP32] [answered]
I am trying to compile the latest commit of Micropython (1.23.0 preview) with IDF 5.2 and Cygwin on Windows 10, every moving part seems to work with the exception of the `qstr.i.last` generator:

```
[1048/1365] cmd.exe /C "cd /D D:\cardputer\micropython\ports\esp32\build-ESP32_GENERIC_S3\esp-idf\main_esp32s3 ...icropython/py/makeversionhdr.py D:/cardputer/micropython/ports/esp32/build-ESP32_GENERIC_S3/genhdr/mpversion.h 
GEN D:/cardputer/micropython/ports/esp32/build-ESP32_GENERIC_S3/genhdr/mpversion.h
[1051/1365] Generating ../../genhdr/qstr.i.last
FAILED: genhdr/qstr.i.last D:/cardputer/micropython/ports/esp32/build-ESP32_GENERIC_S3/genhdr/qstr.i.last
esp-idf\main_esp32s3\CMakeFiles\qstr.i.last-417609e.bat 4d9603dc957d6037
The system cannot execute the specified program.
Batch file failed at line 3 with errorcode 1
```

I found the file `ports\esp32\build-ESP32_GENERIC_S3\esp-idf\main_esp32s3\CMakeFiles\qstr.i.last-417609e.bat` to see why it was failing, and it turns out the abundance of `CFLAGS` that get dumped into said line 3 make it grow to have 64719 characters, which exceeds the CMD (since it's a `.bat`) limit of 8191 characters, even PowerShell's limit of 32764. Is this normal?

**Comment by stinos** [comment]
Similar problems have occurred a number of times, but I don't think there has been a fix. See for instance https://github.com/micropython/micropython/issues/7012.

**Reply by Lana-chan** [reply]
so it seems... for now I was forced to set up a toolchain from scratch in WSL2 and that works for me, but it's a lousy workaround that'll probably satisfy people enough that fixing the build for windows is low priority...

**Comment by andrewleech** [comment]
To be honest wsl2 has been such a success I don't know anyone using cygwin at all anymore, WSL is just so much more stable and easy to work with.

You can also go over step further and use docker (in WSL) to perform all the compilation inside the official Espressif docker containers to avoid having to install or configure a toolchain. It also makes it easy to switch between toolchain versions.

For example, when run from a fresh clone  micropython folder, this will compile and flash an esp board:
```
docker run -it --rm -e BOARD=ESP32_GENERIC -e UID=$(id -u) -v $(pwd):/micropython -v /sys/bus:/sys/bus -v /dev:/dev --net=host --privileged -w /micropython espressif/idf:v5.1.2 bash -c '\
useradd -ms /bin/bash -g root -G sudo -u $UID esp;\
git config --global --add safe.directory /opt/esp/idf;\
su esp -m -c ". /opt/esp/entrypoint.sh; export HOME=/tmp; \
make -C mpy-cross && \
make -C ports/esp32 BOARD=$BOARD submodules && \
make -C ports/esp32 BOARD=$BOARD deploy PORT=/dev/ttyUSB0; \
"'
```
The only things needed beforehand from windows is wsl2, docker desktop installed and the USB device attached to WSL (https://gitlab.com/alelec/wsl-usb-gui/-/releases)

**Comment by stinos** [comment, marked as answer]
Also if not WSL I'd use MSYS2 over Cygwin, I seem to have less problems with it.

**Comment by mzakharo** [comment]
This fix: #17226 is working for me as a solution for building ESP32 Natively on Windows.
